mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala

```diff
@@ -0,0 +1,333 @@
+  setDefault(solver -> AdamW)
+
+  override protected[spark] def train(dataset: Dataset[_]): FMClassifierModel = {
+    val handlePersistence = dataset.storageLevel == StorageLevel.NONE
+    train(dataset, handlePersistence)
+  }
+
+  protected[spark] def train(
+      dataset: Dataset[_],
+      handlePersistence: Boolean): FMClassifierModel = instrumented { instr =>
+    val data: RDD[(Double, OldVector)] =
+      dataset.select(col($(labelCol)), col($(featuresCol))).rdd.map {
+        case Row(label: Double, features: Vector) =>
+          require(label == 0 || label == 1, s"FMClassifier was given" +
+            s" dataset with invalid label $label.  Labels must be in {0,1}; note that" +
+            s" FMClassifier currently only supports binary classification.")
+          (label, features)
+      }
+
+    if (handlePersistence) data.persist(StorageLevel.MEMORY_AND_DISK)
+
+    val numClasses = 2
+    if (isDefined(thresholds)) {
+      require($(thresholds).length == numClasses, this.getClass.getSimpleName +
+        ".train() called with non-matching numClasses and thresholds.length." +
+        s" numClasses=$numClasses, but thresholds has length ${$(thresholds).length}")
+    }
+
+    instr.logPipelineStage(this)
+    instr.logDataset(dataset)
+    instr.logParams(this, factorSize, fitBias, fitLinear, regParam,
+      miniBatchFraction, initStd, maxIter, stepSize, tol, solver)
+    instr.logNumClasses(numClasses)
+
+    val numFeatures = data.first()._2.size
+    instr.logNumFeatures(numFeatures)
+
+    val coefficients = trainImpl(data, numFeatures, LogisticLoss)
+
+    val (bias, linear, factors) = splitCoefficients(
+      coefficients, numFeatures, $(factorSize), $(fitBias), $(fitLinear))
+
+    if (handlePersistence) data.unpersist()
+
+    copyValues(new FMClassifierModel(uid, bias, linear, factors))
+  }
+
+  @Since("3.0.0")
+  override def copy(extra: ParamMap): FMClassifier = defaultCopy(extra)
+}
+
+@Since("3.0.0")
+object FMClassifier extends DefaultParamsReadable[FMClassifier] {
+
+  @Since("3.0.0")
+  override def load(path: String): FMClassifier = super.load(path)
+}
+
+/**
+ * Model produced by [[FMClassifier]]
+ */
+@Since("3.0.0")
+class FMClassifierModel private[classification] (
+  @Since("3.0.0") override val uid: String,
+  @Since("3.0.0") val bias: Double,
+  @Since("3.0.0") val linear: Vector,
+  @Since("3.0.0") val factors: Matrix)
+  extends ProbabilisticClassificationModel[Vector, FMClassifierModel]
+    with FMClassifierParams with MLWritable {
+
+  @Since("3.0.0")
+  override val numClasses: Int = 2
+
+  @Since("3.0.0")
+  override val numFeatures: Int = linear.size
+
+  @transient private lazy val oldCoefficients: OldVector =
+    combineCoefficients(bias, linear, factors, $(fitBias), $(fitLinear))
+
+  @transient private lazy val gradient = parseLoss(
+    LogisticLoss, $(factorSize), $(fitBias), $(fitLinear), numFeatures)
+
+  override protected def predictRaw(features: Vector): Vector = {
```

Comment:
I know it will impl the prediction function twice, but impl it in the .ml side will avoid
1, vector conversion from .ml to .mllib
2, having two model coefficients in memory, the FMModel is usually much larger than other linear models.